cmd/minikube/cmd/config/kubernetes_version.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2022 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"context"
+	"sort"
+
+	"github.com/google/go-github/v43/github"
+	"golang.org/x/mod/semver"
+	"k8s.io/minikube/pkg/minikube/constants"
+)
+
+// supportedKubernetesVersions returns reverse-sort supported Kubernetes releases from GitHub that are in [constants.OldestKubernetesVersion, constants.NewestKubernetesVersion] range, including prereleases.
+// in case it cannot get it from GitHub, in addition to [constants.NewestKubernetesVersion, constants.OldestKubernetesVersion], 'constants.DefaultKubernetesVersion' is also returned if different from 'constants.NewestKubernetesVersion'.
+func supportedKubernetesVersions() (releases []string) {
+	minver := constants.OldestKubernetesVersion
+	defver := constants.DefaultKubernetesVersion
+	maxver := constants.NewestKubernetesVersion
+
+	ghc := github.NewClient(nil)
+
+	opts := &github.ListOptions{PerPage: 100}
+	for (opts.Page+1)*100 <= 300 {
+		rls, resp, err := ghc.Repositories.ListReleases(context.Background(), "kubernetes", "kubernetes", opts)
```

Comment:
@prezha one of the use cases of this feature would be in the minikube GUI, when we load the page to create a cluster we can list all supported kubernetes versions, this will make a http call (not very cheap) in the gui.

i suggest this code to only reference the constants and if needed add some sort of automation (in Makefile) to generate this list. instead of making the call each single time.

is that reasonable?

Reply:
@medyagh that's a good use case as well

this func (exposed via CLI) takes ~1.7sec to produce a fresh list of all currently acceptable k8s versions, which indeed is not that cheap

now, the release cadence for minikube is ~monthly, whereas k8s (supporting the last 3 of its minor versions) has it more frequently, and therefore we/user could end up with the outdated static list (especially if a user did not upgrade to the latest minikube version)

i'd suggest keeping this func/cli functionality as-is in this pr, and for the GUI use case, we could ship minikube releases with a "hardcoded" list of all supported k8s versions (known at the time of the release) and also have eg, a "Refresh" link/button nearby to call this func again so that the GUI can regenerate and cache results for later reuse?

what do you think?

Reply:
As the person working on the GUI, on GUI startup we could make a call this function and cache it. So by the time the user gets around to starting the cluster with a custom k8s version, the list will already be cached and will be able to validate the users input with no latency. I'm not arguing the solution either way, but stating that I don't believe this will impact the GUI.

Reply:
@spowelljr yes, thanks, that sounds like a good/optimal solution to me

@medyagh would you agree then that we keep this func[tionality] as-is in this pr and Steven will call it as needed and also cache results on the gui side?